In [ ]:
import pandas as pd
import numpy as np

In [ ]:
from google.colab import drive
drive.mount('/content/drive')

Mounted at /content/drive


In [ ]:
mkt = pd.read_excel('/content/drive/MyDrive/WeGrowTest/Database - Q3_2020.xlsx', sheet_name = 'MKT')
sales = pd.read_excel('/content/drive/MyDrive/WeGrowTest/Database - Q3_2020.xlsx', sheet_name = 'Sales')
vd = pd.read_excel('/content/drive/MyDrive/WeGrowTest/Database - Q3_2020.xlsx', sheet_name = 'Vận đơn')

# **Phần 1: Xử lý dữ liệu**

## **1- Bảng MKT**

In [ ]:
#Tạo bản sao bảng
mkt_1 = mkt.copy()

In [ ]:
#Shape
num_rows, num_cols = mkt_1.shape
print('Số dòng: ',num_rows)
print('Số cột: ',num_cols)

Số dòng:  1511
Số cột:  21


In [ ]:
mkt_1.head()

Date Channel    MKTer Chiến dịch  Chi phí Marketing  Impression  \
0 2020-07-01      FB    DucVu        CVS                  0           0   
1 2020-07-01      FB    DucVu       Mess                  0           0   
2 2020-07-01      FB    DucVu      Brand             259352       61673   
3 2020-07-01      FB  NgocNgo        CVS            4541746       95958   
4 2020-07-01      FB  NgocNgo      Brand                  0           0   

     Reach  Click  Share  Cmt  ... Lead MKT  Đơn hàng     Doanh thu  \
0      0.0      0      0    0  ...        1   1.00000  4.990000e+05   
1      0.0      0      0    0  ...        3   1.00000  4.990000e+05   
2  59968.0    271      0    0  ...        0   0.00000  0.000000e+00   
3  76992.0    386      4   26  ...       40  31.73913  1.583783e+07   
4      0.0      0      0    0  ...        0   0.00000  0.000000e+00   

   Paid Revenue 1   Giá Lead  Đơn/Lead           CPM           CPC  \
0          499000       0.00  1.000000      0.000000      0.000000   
1          499000       0.00  0.333333      0.000000      0.000000   
2               0       0.00  0.000000   4205.276215    957.018450   
3        12974000  113543.65  0.793478  47330.561287  11766.181347   
4               0       0.00  0.000000      0.000000      0.000000   

   Giá Mess\n(Cmt + Inbox)  ME1 (ME đã bao gồm thuế + phí)  
0                 0.000000                            0.00  
1                 0.000000                            0.00  
2                 0.000000                       280100.16  
3            174682.538462                      4905085.68  
4                 0.000000                            0.00  

[5 rows x 21 columns]

In [ ]:
mkt_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1511 entries, 0 to 1510
Data columns (total 21 columns):
 #   Column                          Non-Null Count  Dtype         
---  ------                          --------------  -----         
 0   Date                            1511 non-null   datetime64[ns]
 1   Channel                         1511 non-null   object        
 2   MKTer                           1511 non-null   object        
 3   Chiến dịch                      1511 non-null   object        
 4   Chi phí Marketing               1511 non-null   int64         
 5   Impression                      1511 non-null   int64         
 6   Reach                           1511 non-null   float64       
 7   Click                           1511 non-null   int64         
 8   Share                           1511 non-null   int64         
 9   Cmt                             1511 non-null   int64         
 10  Inbox                           1511 non-null   object        
 11  Lead

**Xử lí giá trị null**

In [ ]:
mkt_1.isna().sum()

Date                               0
Channel                            0
MKTer                              0
Chiến dịch                         0
Chi phí Marketing                  0
Impression                         0
Reach                              0
Click                              0
Share                              0
Cmt                                0
Inbox                              0
Lead MKT                           0
Đơn hàng                           0
Doanh thu                          0
Paid Revenue 1                     0
Giá Lead                           0
Đơn/Lead                           0
CPM                                0
CPC                                0
Giá Mess\n(Cmt + Inbox)            0
ME1 (ME đã bao gồm thuế + phí)    14
dtype: int64

Cột ME1 xuất hiện 14 giá trị null. Phần dữ liệu này chiếm tỉ lệ nhỏ, hơn nữa chưa có thông tin về thuế nên rất không thể tính toán, nên quyết định xóa những giá trị null này.

In [ ]:
#Xóa giá trị null
mkt_1 = mkt_1.dropna(axis = 1)

**Check duplicate**





In [ ]:
#Check duplicate
duplicated_data = mkt_1[mkt_1.duplicated()]
len(duplicated_data)

1

In [ ]:
duplicated_data = mkt_1[mkt_1.duplicated()]
duplicated_data

Date Channel  MKTer Chiến dịch  Chi phí Marketing  Impression  \
636 2020-08-05      FB  DucVu       Mess                  0           0   

     Reach  Click  Share  Cmt Inbox  Lead MKT  Đơn hàng  Doanh thu  \
636    0.0      0      0    0     0         0       0.0        0.0   

     Paid Revenue 1  Giá Lead  Đơn/Lead  CPM  CPC  Giá Mess\n(Cmt + Inbox)  
636               0       0.0       0.0  0.0  0.0                      0.0

Xóa giá trị duplicates

In [ ]:
#Xóa giá trị dup
mkt_1 = mkt_1.drop_duplicates()

**Kiểm tra lại kiểu dữ liệu**

In [ ]:
mkt_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 1510 entries, 0 to 1510
Data columns (total 20 columns):
 #   Column                  Non-Null Count  Dtype         
---  ------                  --------------  -----         
 0   Date                    1510 non-null   datetime64[ns]
 1   Channel                 1510 non-null   object        
 2   MKTer                   1510 non-null   object        
 3   Chiến dịch              1510 non-null   object        
 4   Chi phí Marketing       1510 non-null   int64         
 5   Impression              1510 non-null   int64         
 6   Reach                   1510 non-null   int64         
 7   Click                   1510 non-null   int64         
 8   Share                   1510 non-null   int64         
 9   Cmt                     1510 non-null   int64         
 10  Inbox                   1510 non-null   object        
 11  Lead MKT                1510 non-null   int64         
 12  Đơn hàng                1510 non-null   float64 

In [ ]:
#Đổi lại kiểu dữ liệu một số cột để phù hợp về mặt ý nghĩa
#Đơn hàng
mkt_1['Đơn hàng'] = round(mkt_1['Đơn hàng'], 0)
mkt_1['Đơn hàng'] = mkt_1['Đơn hàng'].astype(int)
#Đơn/Lead
mkt_1['Đơn/Lead'] = round(mkt_1['Đơn/Lead'], 0)
mkt_1['Đơn/Lead'] = mkt_1['Đơn/Lead'].astype(int)
#Inbox
mkt_1['Inbox'] = mkt_1['Inbox'].astype(int)

In [ ]:
#Kiểm tra lại cột Inbox do không đổi được kiểu dữ liệU
mkt_1['Inbox'].unique()

array(['0', '1', '18', '-'], dtype=object)

In [ ]:
mkt_1[mkt_1['Inbox'] == '-']

Date Channel      MKTer Chiến dịch  Chi phí Marketing  Impression  \
1260  2020-09-14      FB  GiangPhan        CVS            2021614       31348   

      Reach  Click  Share  Cmt Inbox  Lead MKT  Đơn hàng     Doanh thu  \
1260  28080    130      5    9     -        15        15  7.463304e+06   

      Paid Revenue 1       Giá Lead  Đơn/Lead           CPM           CPC  \
1260         6986000  134774.266667         1  64489.409213  15550.876923   

      Giá Mess\n(Cmt + Inbox)  
1260                      0.0

In [ ]:
#Xóa bản ghi có giá trị '-' ở cột Inbox
mkt_1 = mkt_1.drop(mkt_1[mkt_1['Inbox'] == '-'].index)

**Kiểm tra lại giá trị null**

In [ ]:
mkt_1.isna().sum()

Date                       0
Channel                    0
MKTer                      0
Chiến dịch                 0
Chi phí Marketing          0
Impression                 0
Reach                      0
Click                      0
Share                      0
Cmt                        0
Inbox                      0
Lead MKT                   0
Đơn hàng                   0
Doanh thu                  0
Paid Revenue 1             0
Giá Lead                   0
Đơn/Lead                   0
CPM                        0
CPC                        0
Giá Mess\n(Cmt + Inbox)    0
dtype: int64

In [ ]:
#Lưu file
mkt_1.to_excel('MKT_cleaned.xlsx', index = False)


## **2- Bảng Sales**

In [ ]:
#Tạo bản sao bảng sales
sales_1 = sales.copy()

In [ ]:
sales_1.head()

Unnamed: 0                 Giờ        Khách hàng          SĐT Channel  \
0   1-7-2020 2023-07-01 00:13:42     Pham thu thuy  123400000.0      FB   
1   1-7-2020 2023-07-01 00:17:53  Hoàng Thị Lương   123400001.0      FB   
2   1-7-2020 2023-07-01 02:16:56    Phạm Thị huong  123400002.0      FB   
3   1-7-2020 2023-07-01 05:21:32               Lan  123400003.0      FB   
4   1-7-2020 2023-07-01 05:21:31       Lại Thị Đua  123400004.0      FB   

  Chiến dịch     Content   Marketer 2 Type of Lead  \
0        CVS  B1-1812-P1  AnhNguyen08      Dathang   
1        CVS         cay      NgocNgo      Dathang   
2        CVS      matcon      NgocNgo      Dathang   
3        CVS         cay      NgocNgo      Dathang   
4        CVS  condau_sto  YenNguyen02      Dathang   

  Sales Admin xác nhận Type of Lead     Sales  Số lần tương tác   Ngày gọi  \
0                          DATTHANG  ThuyTran               1.0 2020-07-01   
1                          DATTHANG  ThuyTran               1.0 2020-07-01   
2                          DATTHANG  ThuyTran               1.0 2020-07-03   
3                          DATTHANG  ThuyTran               1.0 2020-07-03   
4                          DATTHANG  ThuyTran               1.0 2020-07-03   

             Trạng thái                               Level Ngày hẹn gọi lại  \
0  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
1              Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   
2  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
3  Hủy đơn sau khi chốt         L7.2 - Hủy đơn sau khi chốt              NaT   
4              Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ              NaT   

  Close Date     Tỉnh/TP  Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
0        NaT         NaN               1.0          140000.0     499000  
1 2023-07-04  Quảng Ninh               1.0          140000.0     499000  
2        NaT         NaN               1.0          140000.0     499000  
3        NaT         NaN               1.0          140000.0     499000  
4 2023-07-05      Hà Nội               1.0          140000.0     499000

In [ ]:
num_rows, num_cols = sales_1.shape
print('Số dòng: ',num_rows)
print('Số cột: ',num_cols)

Số dòng:  9675
Số cột:  21


In [ ]:
sales_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 9675 entries, 0 to 9674
Data columns (total 21 columns):
 #   Column                             Non-Null Count  Dtype         
---  ------                             --------------  -----         
 0   Unnamed: 0                         9504 non-null   object        
 1   Giờ                                9504 non-null   datetime64[ns]
 2   Khách hàng                         9504 non-null   object        
 3   SĐT                                9504 non-null   float64       
 4   Channel                            9504 non-null   object        
 5   Chiến dịch                         9504 non-null   object        
 6   Content                            9500 non-null   object        
 7   Marketer 2                         6781 non-null   object        
 8   Type of Lead                       9498 non-null   object        
 9   Sales Admin xác nhận Type of Lead  5448 non-null   object        
 10  Sales                              9

In [ ]:
sales_1.isna().sum()

Unnamed: 0                            171
Giờ                                   171
Khách hàng                            171
SĐT                                   171
Channel                               171
Chiến dịch                            171
Content                               175
Marketer 2                           2894
Type of Lead                          177
Sales Admin xác nhận Type of Lead    4227
Sales                                 171
Số lần tương tác                      179
Ngày gọi                             9471
Trạng thái                            357
Level                                1691
Ngày hẹn gọi lại                     9627
Close Date                           3514
Tỉnh/TP                              6281
Số lượng bộ sách                     2624
Số tiền giảm giá                     2633
Tổng tiền                               0
dtype: int64

Qua quan sát các thông tin data, nhận thấy một số vấn đề sau:


*   Cột đầu tiên của bảng đang không có tên và sai kiểu dữ liệu
*   Trong cột 'Giờ' đang xuất hiện giá trị date

*   Cột 'SĐT' đang có kiểu dữ liệu là float
*   Cột 'Giờ' và 'Closed Date' đang có giá trị năm là 2023 trong khi đây là dữ liệu của năm 2020

*   Data có nhiều giá trị null ở nhiều cột khác nhau








***Đối với cột không tên***, cột này chứa các giá trị ngày tháng, cộng với thông tin của data, có thể suy ra đây là ngày công ty sử dụng các chiến dịch Marketing cho các khách hàng của mình.

***Đổi tên và kiểu dữ liệu cho cột không tên***





In [ ]:
#Đổi kiểu dữ liệu sang datetime
sales_1['Unnamed: 0'] = pd.to_datetime(sales_1['Unnamed: 0'], dayfirst = True)
#Đổi tên cột 'Unnamed: 0' thành 'Ngày'
sales_1 = sales_1.rename(columns = {'Unnamed: 0': 'Ngày'})

<ipython-input-28-367b7518b0a8>:3: UserWarning: Parsing dates in DD/MM/YYYY format when dayfirst=False (the default) was specified. This may lead to inconsistently parsed dates! Specify a format to ensure consistent parsing.
  sales_1['Unnamed: 0'] = pd.to_datetime(sales_1['Unnamed: 0'])


***Xóa phần date của cột 'Giờ'***




In [ ]:
sales_1['Giờ'] = sales_1['Giờ'].dt.time


***Sửa lại giá trị năm của cột 'Closed Date' thành 2020.***

In [ ]:
#Kiểm tra các giá trị năm ở cột này
sales_1['Năm'] = sales_1['Close Date'].dt.year
sales_1['Năm'].value_counts()

2023.0    6161
Name: Năm, dtype: int64

Có thể thấy, giống với cột 'Giờ' ban đầu, cột 'Close Date' cũng bị nhập lỗi giá trị năm là 2023 và cần sửa lại.

In [ ]:
#Xóa cột 'Năm'
del sales_1['Năm']
#Thay giá trị năm thành 2020
sales_1['Close Date'] = sales_1['Close Date'].apply(lambda x: x.replace(year=2020))

**Xử lý các giá trị null**

In [ ]:
#CheckNull
sales_1.isna().sum()

Ngày                                  171
Giờ                                   171
Khách hàng                            171
SĐT                                   171
Channel                               171
Chiến dịch                            171
Content                               175
Marketer 2                           2894
Type of Lead                          177
Sales Admin xác nhận Type of Lead    4227
Sales                                 171
Số lần tương tác                      179
Ngày gọi                             9471
Trạng thái                            357
Level                                1691
Ngày hẹn gọi lại                     9627
Close Date                           3514
Tỉnh/TP                              6281
Số lượng bộ sách                     2624
Số tiền giảm giá                     2633
Tổng tiền                               0
dtype: int64

Nhận thấy số giá trị null ở 7 dòng đều bằng 171, nên thử kiểm tra vùng dữ liệu này.

In [ ]:
#Xem xét vùng dữ liệu khi cột 'Ngày' null
ngay_null = sales_1[sales_1['Ngày'].isna()]
ngay_null.isna().sum()

Ngày                                 171
Giờ                                  171
Khách hàng                           171
SĐT                                  171
Channel                              171
Chiến dịch                           171
Content                              171
Marketer 2                           171
Type of Lead                         171
Sales Admin xác nhận Type of Lead    171
Sales                                171
Số lần tương tác                     171
Ngày gọi                             171
Trạng thái                           171
Level                                171
Ngày hẹn gọi lại                     171
Close Date                           171
Tỉnh/TP                              171
Số lượng bộ sách                     171
Số tiền giảm giá                     171
Tổng tiền                              0
dtype: int64

Nhận thấy toàn bộ vùng dữ liệu này đều có giá trị null, nên loại bỏ nó khỏi dataframe.

In [ ]:
#Xóa bỏ vùng dữ liệu
sales_1 = sales_1.drop(sales_1[sales_1['Ngày'].isna()].index)

In [ ]:
#Tiếp tục kiểm tra giá trị null
sales_1.isna().sum()

Ngày                                    0
Giờ                                     0
Khách hàng                              0
SĐT                                     0
Channel                                 0
Chiến dịch                              0
Content                                 4
Marketer 2                           2723
Type of Lead                            6
Sales Admin xác nhận Type of Lead    4056
Sales                                   0
Số lần tương tác                        8
Ngày gọi                             9300
Trạng thái                            186
Level                                1520
Ngày hẹn gọi lại                     9456
Close Date                           3343
Tỉnh/TP                              6110
Số lượng bộ sách                     2453
Số tiền giảm giá                     2462
Tổng tiền                               0
dtype: int64

Xóa một số cột sau:

*   'Sales Admin xác nhận Type of Lead': đây là cột chứa thông xác nhận của Sales Admin, có giá trị gần như tương đồng với cột 'Type of Lead'.
*   'Ngày gọi', 'Ngày hẹn gọi lại': 2 cột gần chứa gần như đều là giá trị null.


In [ ]:
#Xóa cột
sales_1 = sales_1.drop(columns=['Sales Admin xác nhận Type of Lead','Ngày gọi','Ngày hẹn gọi lại'])

Kiểm tra các cột có giá trị null còn lại, gồm 9 cột:



*  Content

*   Marketer

*   Type of Lead

*   Số lần tương tác
*   Trạng thái


*  Level


*   Close Date


*   Tỉnh/TP


*   Số lượng bộ sách


*   Số tiền giảm giá







1- Cột Content



In [ ]:
sales_1[sales_1['Content'].isna()]

Ngày       Giờ  Khách hàng          SĐT Channel Chiến dịch Content  \
4228 2020-09-23  10:43:08       Hường  123404228.0      FB        CVS     NaN   
6004 2020-07-29  12:20:17   Nguyệt Vũ  123406002.0      FB        CVS     NaN   
8175 2020-06-08  13:39:25  Thùy Khiêm  123408173.0      FB        CVS     NaN   
9114 2020-09-07  06:22:28     Bui Nga  123409112.0      FB        CVS     NaN   

       Marketer 2 Type of Lead Sales Admin xác nhận Type of Lead  ...  \
4228          NaN      Dathang                               NaN  ...   
6004  YenNguyen02      Dathang                          DATTHANG  ...   
8175  YenNguyen02        Tuvan                               NaN  ...   
9114    GiangPhan      Dathang                          DATTHANG  ...   

     Số lần tương tác  Ngày gọi Trạng thái  \
4228              2.0       NaT   Đặt hàng   
6004              1.0       NaT   Đặt hàng   
8175              5.0       NaT   Đặt hàng   
9114              1.0       NaT   Đặt hàng   

                                   Level Ngày hẹn gọi lại Close Date  \
4228  L8.2 - Khách đã thanh toán toàn bộ              NaT 2020-09-27   
6004  L8.2 - Khách đã thanh toán toàn bộ              NaT 2020-08-01   
8175             L7.1 - Chờ chuyển khoản              NaT        NaT   
9114                                 NaN              NaT 2020-07-11   

         Tỉnh/TP Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
4228         NaN              1.0          140000.0     499000  
6004    Bắc Ninh              1.0          140000.0     499000  
8175  Bình Dương              1.0          240000.0     399000  
9114      Hà Nội              1.0          140000.0     499000  

[4 rows x 21 columns]

In [ ]:
sales_1['Content'].value_counts()

lop11            2577
digheconchong     524
condau_sto        503
ac                423
haoi              394
                 ... 
tonthuong           1
100baihocanh1       1
vietkieu_an         1
chidau              1
vidkol              1
Name: Content, Length: 353, dtype: int64

Nhận thấy cột 'Content' có rất nhiều giá trị riêng biệt, cộng thêm chưa tìm được quy luật của các giá trị trong cột, nên quyết định xóa các giá trị này.

In [ ]:
sales_1 = sales_1.drop(sales_1[sales_1['Content'].isna()].index)

2- Cột 'Marketer 2'

In [ ]:
sales_1['Marketer 2'].value_counts(dropna=False)

NaN              2723
DucVu            1851
YenNguyen02      1748
GiangPhan        1013
NgocNgo           836
TienTran          517
TungTran          512
AnhNguyen08       195
ThuHoang           48
NhiLanh            21
AnhTrinh           11
LinhHoang          10
HuyDo               7
TrangDang           3
LyLuu               3
TuanTran            2
AnhNgo01            1
ht.content.05       1
HuyenNguyen05       1
                    1
Name: Marketer 2, dtype: int64

Cột chứa rất nhiều giá trị khác nhau và không có quy luật nhất định, nên giữ nguyên những giá trị null

3- Cột 'Số lần tương tác'

In [ ]:
sales_1[sales_1['Số lần tương tác'].isna()]

Ngày       Giờ        Khách hàng          SĐT Channel Chiến dịch  \
204  2020-01-09  14:49:18        Thanh Pham  123400204.0      FB        CVS   
522  2020-10-09  08:38:34        Tran Nhung  123400522.0      FB        CVS   
1865 2020-08-15  15:40:03       Lịch Nguyễn  123401865.0      FB        CVS   
2671 2020-09-17  17:16:32  Lưu Thị Mỹ Huyền  123402671.0      FB        CVS   
3369 2020-02-09  13:46:58           Mai Anh  123403369.0      FB        CVS   
5994 2020-09-28  21:30:09   Nguyễn Mạnh Hải  123405992.0  Upsell     Upsell   
6908 2020-08-30  14:30:12       Yumi Nguyen  123406906.0      FB        CVS   
7068 2020-09-30  15:50:27          Hoan Hao  123407066.0      FB        CVS   

               Content Marketer 2 Type of Lead          Sales  \
204              lop11      DucVu        Tuvan       ThaoKieu   
522   page Việt Phương        NaN        Tuvan    AnhNguyen13   
1865     digheconchong    NgocNgo        Tuvan    AnhNguyen13   
2671       dungpthai-2        NaN        Tuvan         NgocDo   
3369      bosachxizach   TungTran        Tuvan    AnhNguyen13   
5994            Upsell        NaN      Dathang     NganNguyen   
6908             lop11    NgocNgo        Tuvan    HuyenNguyen   
7068     ht.content.05        NaN        Tuvan  HuyenNguyen04   

      Số lần tương tác Trạng thái                               Level  \
204                NaN        NaN                                 NaN   
522                NaN        NaN                                 NaN   
1865               NaN        NaN                                 NaN   
2671               NaN      Trùng                                 NaN   
3369               NaN        NaN                                 NaN   
5994               NaN   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
6908               NaN        NaN                                 NaN   
7068               NaN        NaN                                 NaN   

     Close Date Tỉnh/TP  Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
204         NaT     NaN               NaN               NaN          0  
522         NaT     NaN               NaN               NaN          0  
1865        NaT     NaN               NaN               NaN          0  
2671        NaT     NaN               NaN               NaN          0  
3369        NaT     NaN               NaN               NaN          0  
5994 2020-09-30     NaN               1.0          140000.0     499000  
6908        NaT     NaN               NaN               NaN          0  
7068        NaT     NaN               NaN               NaN          0

In [ ]:
sales_1['Số lần tương tác'].value_counts(dropna=False)

1.0     7089
2.0      767
3.0      600
4.0      346
5.0      227
6.0      220
7.0      137
8.0       73
9.0       24
10.0       9
NaN        8
11.0       4
Name: Số lần tương tác, dtype: int64

Cột có nhiều giá trị riêng biệt và không có quy luật nên khá khó để điền những giá trị null này, ngoài ra số giá trị null chiếm tỉ trọng nhỏ nên quyết định xóa những bản ghi này.

In [ ]:
#Xóa giá trị null
sales_1 = sales_1.drop(sales_1[sales_1['Số lần tương tác'].isna()].index)

4- Cột 'Type of Lead'

In [ ]:
sales_1[sales_1['Type of Lead'].isna()]

Ngày       Giờ        Khách hàng          SĐT    Channel  \
124  2020-01-08  15:55:45        Minh Trang  123400124.0         FB   
695  2020-11-07  15:23:21         Chị Tuyến  123400695.0         FB   
1393 2020-08-13  09:12:36  Nguyễn Thu Giang  123401393.0  Saleshunt   
3171 2020-09-19  15:44:19          chị Hiền  123403171.0  Saleshunt   
5437 2020-07-27  09:33:42         Thanh Tâm  123405437.0         FB   
7140 2020-07-31  15:26:54      Phan Mỹ Linh  123407138.0     Upsell   

     Chiến dịch           Content Marketer 2 Type of Lead          Sales  \
124        Mess  leadkhongronguon      DucVu          NaN          AnhHa   
695         CVS           hotline      DucVu          NaN         DatLam   
1393  Saleshunt         Saleshunt  LinhHoang          NaN  HuyenNguyen04   
3171  Saleshunt         Saleshunt        NaN          NaN  HuyenNguyen04   
5437       Mess  leadkhongronguon      DucVu          NaN  HuyenNguyen04   
7140     Upsell            Upsell  LinhHoang          NaN        HoaTran   

      Số lần tương tác Trạng thái                               Level  \
124                1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
695                1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
1393               1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
3171               1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
5437               1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   
7140               1.0   Đặt hàng  L8.2 - Khách đã thanh toán toàn bộ   

     Close Date    Tỉnh/TP  Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
124  2020-08-04     Hà Nội               1.0         -160000.0     799000  
695  2020-07-17   Bắc Ninh               1.0          140000.0     499000  
1393 2020-08-13     Hà Nội               1.0          140000.0     499000  
3171 2020-09-21  Hải Dương               1.0          140000.0     499000  
5437 2020-07-30    Đà Nẵng               1.0          140000.0     499000  
7140 2020-07-28        NaN               1.0          359000.0     280000

In [ ]:
sales_1['Type of Lead'].value_counts(dropna=False)

Dathang               6232
Tuvan                 3257
NaN                      6
Con đang học lớp 7       1
Name: Type of Lead, dtype: int64

Tương tự như cột 'Số lần tương tác', loại bỏ các bản ghi có giá trị null. Ngoài ra loại bỏ thêm bản ghi có giá trị 'Con đang học lớp 7' vì nó là giá trị ngoại lai.

In [ ]:
#Xóa các dòng null và ngoại lai
sales_1 = sales_1.drop(sales_1[(sales_1['Type of Lead'].isna()) | (sales_1['Type of Lead'] == 'Con đang học lớp 7')].index)
sales_1['Type of Lead'].value_counts(dropna=False)

Dathang    6232
Tuvan      3257
Name: Type of Lead, dtype: int64

5-Cột 'Trạng thái'

In [ ]:
sales_1[sales_1['Trạng thái'].isna()]

Ngày       Giờ           Khách hàng          SĐT Channel  \
280   2020-10-07  04:29:06              Hai Van  123400282.0      FB   
330   2020-10-07  10:25:33      Nguyễn Thị Thúy  123400332.0      FB   
331   2020-10-07  10:27:06   Nguyễn Thị Anh Đào  123400333.0      FB   
332   2020-10-07  10:33:14    Nguyễn Hồng Nhung  123400334.0      FB   
333   2020-10-07  10:34:27       Nguyen Van Anh  123400335.0      FB   
...          ...       ...                  ...          ...     ...   
9180  2020-09-07  16:35:15         Phạm thị mai  123409193.0      FB   
9181  2020-09-07  16:40:04        Phạm Ngọc Mai  123409194.0      FB   
9182  2020-09-07  16:57:03             Ngọc Để   123409195.0      FB   
9183  2020-09-07  16:58:39  Nguyễn Thị Thu Hiền  123409196.0      FB   
9198  2020-09-07  20:54:23          Đỗ Thị luân  123409211.0      FB   

     Chiến dịch               Content   Marketer 2 Type of Lead  \
280       Brand  docthu_GuidelineBook      NhiLanh      Dathang   
330         CVS            condau_sto  YenNguyen02      Dathang   
331         CVS            condau_sto  YenNguyen02      Dathang   
332         CVS                   bcs  YenNguyen02        Tuvan   
333         CVS            condau_sto  YenNguyen02        Tuvan   
...         ...                   ...          ...          ...   
9180        CVS            condau_sto  YenNguyen02      Dathang   
9181        CVS            condau_sto  YenNguyen02      Dathang   
9182        CVS                  haoi    GiangPhan      Dathang   
9183        CVS            condau_sto  YenNguyen02      Dathang   
9198        CVS           khacbiet-we    GiangPhan      Dathang   

              Sales  Số lần tương tác Trạng thái  \
280          NgocDo               1.0        NaN   
330        ThuyTran               1.0        NaN   
331        ThuyTran               1.0        NaN   
332   HuyenNguyen04               2.0        NaN   
333       HuongDinh               6.0        NaN   
...             ...               ...        ...   
9180       ThuyTran               1.0        NaN   
9181       ThuyTran               1.0        NaN   
9182       ThuyTran               1.0        NaN   
9183       ThuyTran               1.0        NaN   
9198       ThuyTran               1.0        NaN   

                                   Level  Close Date     Tỉnh/TP  \
280   L8.2 - Khách đã thanh toán toàn bộ  2020-07-12      Hà Nội   
330   L8.2 - Khách đã thanh toán toàn bộ  2020-07-13      Hà Nội   
331   L8.2 - Khách đã thanh toán toàn bộ  2020-07-12   Hải Dương   
332                                  NaN         NaN         NaN   
333                                  NaN         NaN         NaN   
...                                  ...         ...         ...   
9180  L8.2 - Khách đã thanh toán toàn bộ  2020-07-13   Khánh Hòa   
9181  L8.2 - Khách đã thanh toán toàn bộ  2020-07-12  Quảng Ninh   
9182  L8.2 - Khách đã thanh toán toàn bộ  2020-07-13  Bình Dương   
9183  L8.2 - Khách đã thanh toán toàn bộ  2020-07-12    Bắc Ninh   
9198  L8.2 - Khách đã thanh toán toàn bộ  2020-07-12      Hà Nội   

      Số lượng bộ sách  Số tiền giảm giá  Tổng tiền  
280                1.0          140000.0     499000  
330                1.0          140000.0     499000  
331                1.0          140000.0     499000  
332                NaN               NaN          0  
333                NaN               NaN          0  
...                ...               ...        ...  
9180               1.0          140000.0     499000  
9181               1.0          140000.0     499000  
9182               1.0          140000.0     499000  
9183               1.0          140000.0     499000  
9198               1.0          140000.0     499000  

[180 rows x 18 columns]

Đối với cột 'Trạng thái', nhận thấy cột này khá tương đồng với cột 'Level' vì cả 2 đều có nội dung liên quan đến trạng thái của khách (đặt hàng, từ chối đặt hàng, hủy đơn,... ), nên sẽ điền các giá trị null của cột 'Trạng thái' dựa trên cột 'Level'.

Trước tiên, cần xử lí cột 'Level'. Có thể thấy, ở đầu mỗi giá trị trong cột có các ký tự gây khó hiểu như: L8.1,L8.2,... Vì vậy cần loại bỏ các ký tự này.

In [ ]:
#Bỏ các ký tự trước dấu '-' của cột Level
sales_1['Level'] = sales_1['Level'].str.split('-').str[-1]
#Xóa kí tự " " ở đầu mỗi giá trị trong cột
sales_1['Level'] = sales_1['Level'].str.strip()

Tiếp đến, kiểm tra vùng cột 'Level' khi cột 'Trạng thái' null.

In [ ]:
#Xem các giá trị ở cột level ở vùng dữ liệu null
trangthai_null = sales_1[sales_1['Trạng thái'].isna()]
trangthai_null['Level'].value_counts()

Khách đã thanh toán toàn bộ    136
Hủy đơn sau khi chốt             7
Không nghe máy/Thuê bao          4
Băn khoăn về giá                 1
Trùng                            1
Đã được tư vấn đầy đủ            1
Không còn nhu cầu                1
Name: Level, dtype: int64

Các giá trị xuất hiện ở cột 'Level' khi 'Trạng thái' null bao gồm:

*   Khách đã thanh toán toàn bộ    


*   Hủy đơn sau khi chốt    
*   Không nghe máy/Thuê bao

*   Băn khoăn về giá

*   Trùng    
*   Đã được tư vấn đầy đủ

*   Không còn nhu cầu     

Tiếp theo, kiểm tra khi 'Level' là một trong những giá trị này, 'Trạng thái' sẽ là những giá trị nào.







         
         
                 
                       
          




In [ ]:
#FilterLevel
values_to_keep = ['Khách đã thanh toán toàn bộ','Hủy đơn sau khi chốt','Không nghe máy/Thuê bao','Băn khoăn về giá','Trùng','Đã được tư vấn đầy đủ','Không còn nhu cầu']
#FilterData
Filter_data = sales_1[sales_1['Level'].isin(values_to_keep)]
Filter_data
#GroupBy
groupby = Filter_data.groupby(['Level','Trạng thái'])['Trạng thái'].count()
print(groupby)


Level                        Trạng thái               
Băn khoăn về giá             Hẹn lúc khác gọi lại            2
                             Không nghe máy                  5
                             Suy nghĩ thêm/Trả lời sau      31
                             Từ chối                        18
Hủy đơn sau khi chốt         Hủy đơn sau khi chốt          530
                             Suy nghĩ thêm/Trả lời sau       1
                             Từ chối                         3
                             Đặt hàng                      128
Khách đã thanh toán toàn bộ  Đặt hàng                     5954
Không còn nhu cầu            Từ chối                        22
Không nghe máy/Thuê bao      Hẹn lúc khác gọi lại           45
                             Không nghe máy                 92
                             Sai số                          2
                             Suy nghĩ thêm/Trả lời sau      25
                             Thuê bao                       20


Điền các giá trị null ở cột 'Trạng thái' bằng giá trị xuất hiện nhiều nhất khi quan sát các cặp giá trị của cột 'Level' và 'Trạng thái.

Ví dụ: khi cột 'Level' là "Hủy đơn sau khi chốt", ta sẽ điền vào cột 'Trạng thái' giá trị "Hủy đơn sau khi chốt" do cặp giá trị này xuất hiện nhiều nhất (530 lần). Tương tự với những cột còn lại.

Level:  Băn khoăn về giá --> Suy nghĩ thêm/Trả lời sau

  Hủy đơn sau khi chốt --> Hủy đơn sau khi chốt

  Khách đã thanh toán toàn bộ --> Đặt hàng

  Không còn nhu cầu --> Từ chối

  Không nghe máy/Thuê bao --> Không nghe máy

  Trùng --> Trùng

  Đã được tư vấn đầy đủ --> Suy nghĩ thêm/Trả lời sau



In [ ]:
#Điền giá trị null
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Băn khoăn về giá'), 'Suy nghĩ thêm/Trả lời sau'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Hủy đơn sau khi chốt'), 'Hủy đơn sau khi chốt'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Khách đã thanh toán toàn bộ'), 'Đặt hàng'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Không còn nhu cầu'), 'Từ chối'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Không nghe máy/Thuê bao'), 'Không nghe máy'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Trùng'), 'Trùng'))
sales_1['Trạng thái'] = sales_1['Trạng thái'].fillna(sales_1['Trạng thái'].where(~(sales_1['Level'] == 'Đã được tư vấn đầy đủ'), 'Suy nghĩ thêm/Trả lời sau'))

In [ ]:
#Kiểm tra các giá trị null còn lại ở cột 'Trạng thái'
sales_1['Trạng thái'].isna().sum()

29

In [ ]:
#Xóa các giá trị null cồn lại
sales_1 = sales_1.drop(sales_1[sales_1['Trạng thái'].isna()].index)


In [ ]:
sales_1.isna().sum()

Ngày                   0
Giờ                    0
Khách hàng             0
SĐT                    0
Channel                0
Chiến dịch             0
Content                4
Marketer 2          2714
Type of Lead           0
Sales                  0
Số lần tương tác       0
Trạng thái             0
Level               1484
Close Date          3309
Tỉnh/TP             6072
Số lượng bộ sách    2419
Số tiền giảm giá    2428
Tổng tiền              0
dtype: int64

In [ ]:
sales_1.shape

(9460, 17)

6- Cột 'Level'

Do cột có giá trị khá tương đồng với cột 'Trạng thái', nên khi làm báo cáo hay phân tích sau này chỉ cần sử dụng một trong hai cột. Do đó, xóa cột 'Level'

In [ ]:
#Xóa cột
del sales_1['Level']

7- Cột 'Close Date'

Qua quan sát, nhận thấy cột Close date có liên quan với cột Trạng thái.

In [ ]:
#Phân tích theo cột 'Trạng thái'
#Vùng dữ liệu khi Close Date null
filter_df = sales_1[sales_1['Close Date'].isna()]
#Các giá trị Trạng thái trong vùng dữ liệu này
filter_df['Trạng thái'].value_counts()

Suy nghĩ thêm/Trả lời sau    1119
Hủy đơn sau khi chốt          540
Từ chối                       431
Không nghe máy                421
Đặt hàng                      320
Hẹn lúc khác gọi lại          292
Thuê bao                       81
Sai số                         49
Trùng                          44
Trêu đùa                       12
Name: Trạng thái, dtype: int64

Có thể thấy, các bản ghi có Close date null đều là của những khách hàng với các trạng thái như: suy nghĩ thêm, trả lời sau, hủy đơn, không nghe máy,... Điều này giải thích rằng những giao dịch này chưa được hoàn thành, dẫn tới việc cột Close date bị bỏ trống.

8-Cột 'Tỉnh/TP'

Các giá trị trong cột này bị null rất nhiều. Đã thử gộp bảng 'sales' với bảng 'Vận đơn' thông qua cột tên khách hàng, tuy nhiên lại có rất nhiều tên khách hàng trùng với nhau, nên không thể lấy được thông tin cột 'Tỉnh/TP' từ bảng 'Vận đơn' để điền vào bảng 'Sales'. Vì vậy, giữ nguyên các giá trị null ở cột này.




9- Cột 'Số lượng bộ sách' và 'Số tiền giảm giá'

Đối với cột 'Số lượng bộ sách', lí do dữ liệu bị bỏ trống có thể là do khách hàng không đặt hàng nên số lượng của đơn hàng bị bỏ trống.

Thử kiểm tra cột Trạng thái khi cột Số lượng bộ sách null.

In [ ]:
#Filter
filter = sales_1[sales_1['Số lượng bộ sách'].isna()]
#Các giá trị của Trạng thái
filter['Trạng thái'].value_counts()

Suy nghĩ thêm/Trả lời sau    1109
Từ chối                       422
Không nghe máy                415
Hẹn lúc khác gọi lại          290
Thuê bao                       80
Sai số                         49
Trùng                          36
Trêu đùa                       12
Đặt hàng                        4
Hủy đơn sau khi chốt            2
Name: Trạng thái, dtype: int64

Đúng như đã dự đoán, các giá trị phần lớn đều là: suy nghĩ thêm, từ chối, thuê bao,...

Tiếp theo kiểm tra cột 'Số tiền giảm giá' khi 'Số lượng bộ sách' null.

In [ ]:
filter['Số tiền giảm giá'].value_counts(dropna=False)

NaN    2419
Name: Số tiền giảm giá, dtype: int64

Có thể thấy khi 'Số lượng bộ sách' null thì 'Số tiền giảm giá' cũng null. Điều này là hợp lý vì nếu không đặt hàng thì sẽ không được giảm giá.

Vì vậy, điền các giá trị null bằng giá trị 0 khi hai cột này đều.

In [ ]:
#Fillna Số lượng bộ sách
sales_1['Số lượng bộ sách'] = sales_1['Số lượng bộ sách'].fillna(0)
#Fillna So tiền giảm giá
con = (sales_1['Số lượng bộ sách'] == 0) & (sales_1['Số tiền giảm giá'].isna())
sales_1.loc[con, 'Số tiền giảm giá'] = 0

In [ ]:
#Các giá trị null còn lại của cột Số tiền giảm giá
sales_1['Số tiền giảm giá'].isna().sum()

9

In [ ]:
#Xóa 9 dòng null này
sales_1 = sales_1.drop(sales_1[sales_1['Số tiền giảm giá'].isna()].index)

**Chỉnh sửa một số cột**

In [ ]:
#Các giá trị trong cột 'Số  tiền giảm giá'
sales_1['Số tiền giảm giá'].unique()

array([ 1.40000e+05,  0.00000e+00,  2.80000e+05,  2.40000e+05,
       -1.41000e+05, -1.40000e+05,  1.28150e+05,  2.61730e+04,
        1.39000e+05,  1.67470e+05,  9.58500e+04,  1.07251e+05,
        6.29240e+04,  3.91860e+04,  1.47000e+06,  3.89000e+05,
        2.89000e+05,  4.99000e+05, -2.33647e+05, -5.10000e+04,
        1.00000e+04, -1.00000e+04, -3.00200e+05, -2.24841e+05,
        4.20000e+05, -1.50600e+05, -7.51000e+05,  9.70000e+02,
       -6.78010e+04, -2.59000e+05,  4.93860e+04, -3.02400e+05,
       -2.14844e+05, -1.64266e+05, -7.23130e+05,  1.90000e+05,
        4.29870e+04,  1.52475e+05])

*Ở cột 'Số tiền giảm giá'* có xuất hiện giá trị âm, có thể đây lỗi xảy ra khi nhập liệu. Tiến hành chuyển chúng sang giá trị dương.

In [ ]:
#Chuyển type sang dạng str
sales_1['Số tiền giảm giá'] = sales_1['Số tiền giảm giá'].astype(str)
#Chia cột và lấy phần ký tự sau dấu '-'
sales_1['Số tiền giảm giá'] = sales_1['Số tiền giảm giá'].str.split('-').str[-1]
#Chuyển lại type thành float
sales_1['Số tiền giảm giá'] = sales_1['Số tiền giảm giá'].astype(float)

In [ ]:
#Kiểm tra lại kiểu dữ liệu
sales_1.dtypes

Ngày                 object
Giờ                  object
Khách hàng           object
SĐT                 float64
Channel              object
Chiến dịch           object
Content              object
Marketer 2           object
Type of Lead         object
Sales                object
Số lần tương tác    float64
Trạng thái           object
Close Date           object
Tỉnh/TP              object
Số lượng bộ sách    float64
Số tiền giảm giá    float64
Tổng tiền             int64
dtype: object

Đổi lại kiểu dữ liệu một số cột cho phù hợp

In [ ]:
#Ngày sang datetime
sales_1['Ngày'] = pd.to_datetime(sales_1['Ngày'])
#SĐT sang int
sales_1['SĐT'] = sales_1['SĐT'].astype(int)
#Số lần tương tác sang int
sales_1['Số lần tương tác'] = sales_1['Số lần tương tác'].astype(int)
#Close Date sang datetime
sales_1['Close Date'] = pd.to_datetime(sales_1['Close Date'])
#Số lượng bộ sách sang int
sales_1['Số lượng bộ sách'] = sales_1['Số lượng bộ sách'].astype(int)

**Kiểm tra lại kiểu dữ liệu và giá trị null**

In [ ]:
sales_1.isna().sum()

Ngày                   0
Giờ                    0
Khách hàng             0
SĐT                    0
Channel                0
Chiến dịch             0
Content                0
Marketer 2          2707
Type of Lead           0
Sales                  0
Số lần tương tác       0
Trạng thái             0
Close Date          3300
Tỉnh/TP             6068
Số lượng bộ sách       0
Số tiền giảm giá       0
Tổng tiền              0
dtype: int64

In [ ]:
sales_1.info()

<class 'pandas.core.frame.DataFrame'>
Int64Index: 9447 entries, 0 to 9503
Data columns (total 17 columns):
 #   Column            Non-Null Count  Dtype         
---  ------            --------------  -----         
 0   Ngày              9447 non-null   datetime64[ns]
 1   Giờ               9447 non-null   object        
 2   Khách hàng        9447 non-null   object        
 3   SĐT               9447 non-null   int64         
 4   Channel           9447 non-null   object        
 5   Chiến dịch        9447 non-null   object        
 6   Content           9447 non-null   object        
 7   Marketer 2        6740 non-null   object        
 8   Type of Lead      9447 non-null   object        
 9   Sales             9447 non-null   object        
 10  Số lần tương tác  9447 non-null   int64         
 11  Trạng thái        9447 non-null   object        
 12  Close Date        6147 non-null   datetime64[ns]
 13  Tỉnh/TP           3379 non-null   object        
 14  Số lượng bộ sách  9447 n

In [ ]:
#Xóa giá trị duplicates
sales_1 = sales_1.drop_duplicates()

In [ ]:
#Lưu file
sales_1.to_excel('Sales_cleaned.xlsx', index = False)

## **3- Bảng Vận đơn**

In [ ]:
#Tạo bản copy
vd_1 = vd.copy()

In [ ]:
vd_1.head()

STT Mã đơn hàng                                   Ghi chú đơn hàng  \
0  1.0    SON25144                                                NaN   
1  NaN         NaN                                                NaN   
2  2.0    SON24883  TẶNG SÁCH LẺ PHẦN 2 CHO PHỤ HUỲNH\ngiao trong ...   
3  NaN         NaN                                                NaN   
4  3.0    SON24943                          Giao trong giờ hành chính   

  Tags đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn     Mã vận đơn  \
0           NaN  Nguyễn Thị Huyền  Chi nhánh mặc định   Web       G87L96F3   
1           NaN               NaN                 NaN   NaN            NaN   
2           NaN  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web       G87L9VTM   
3           NaN               NaN                 NaN   NaN            NaN   
4           NaN  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web  NSAPOAAAAZPQF   

  Tình trạng gói hàng Trạng thái đối tác  Lý do hủy đơn       Ngày đóng gói  \
0        Đã giao hàng    Giao thành công            NaN 2020-09-16 12:49:54   
1                 NaN                NaN            NaN                 NaT   
2        Đã giao hàng    Giao thành công            NaN 2020-09-16 12:49:13   
3                 NaN                NaN            NaN                 NaT   
4        Đã giao hàng    Giao thành công            NaN 2020-09-15 19:56:20   

   Ngày hẹn giao       Ngày xuất kho      Ngày giao hàng Đối tác giao hàng  \
0            NaN 2020-09-17 11:05:38 2020-09-19 10:20:38      Sapo Express   
1            NaN                 NaT                 NaT               NaN   
2            NaN 2020-09-17 11:05:48 2020-09-21 12:18:59      Sapo Express   
3            NaN                 NaT                 NaT               NaN   
4            NaN 2020-09-15 19:58:37 2020-09-19 12:54:17      Sapo Express   

     Dịch vụ giao hàng  Khối lượng(g)     Kích thước(DxRxC)  Tên người nhận  \
0                Chuẩn          500.0  10.000x10.000x10.000    Mẹ Bống Trần   
1                  NaN            NaN                   NaN             NaN   
2                Chuẩn          500.0  10.000x10.000x10.000         Nhan Le   
3                  NaN            NaN                   NaN             NaN   
4  (Ninja Van - Chuẩn)         1000.0  10.000x10.000x10.000  Vũ Thị Thúy Hà   

  SĐT người nhận                            Địa chỉ giao hàng      Tỉnh/Thành  \
0     0389913230                                    thôn Ngọc        Hưng Yên   
1            NaN                                          NaN             NaN   
2     0907867931         trường mn Phú Thuận đường Phạm Thị B  TP Hồ Chí Minh   
3            NaN                                          NaN             NaN   
4     0916522265  Bệnh viện phụ sản trung ương Triệu Quốc Đạt          Hà Nội   

       Quận/Huyện         Phường xã Trạng thái đối soát  \
0   Huyện Văn Lâm        Xã Lạc Đạo         Đã đối soát   
1             NaN               NaN                 NaN   
2          Quận 7  Phường Phú Thuận         Đã đối soát   
3             NaN               NaN                 NaN   
4  Quận Hoàn Kiếm  Phường Hàng Bông         Đã đối soát   

   Tiền khách phải trả cho đơn  Khách hàng đã trả Hình thức thanh toán  \
0                     499000.0           499000.0                  COD   
1                          NaN                NaN                  NaN   
2                     499000.0           499000.0                  COD   
3                          NaN                NaN                  NaN   
4                     780000.0           780000.0                  COD   

   Tổng tiền thu hộ  Phí vận chuyển Người trả phí  Phí trả đối tác  \
0          499000.0             0.0      Shop trả          26600.0   
1               NaN             NaN           NaN              NaN   
2          499000.0             0.0      Shop trả          26600.0   
3               NaN             NaN           NaN              NaN   
4          780000.0             0.0    

In [ ]:
vd_1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7944 entries, 0 to 7943
Data columns (total 45 columns):
 #   Column                       Non-Null Count  Dtype         
---  ------                       --------------  -----         
 0   STT                          5820 non-null   float64       
 1   Mã đơn hàng                  5820 non-null   object        
 2   Ghi chú đơn hàng             566 non-null    object        
 3   Tags đơn hàng                27 non-null     object        
 4   Nhân viên tạo đơn            5820 non-null   object        
 5   Chi nhánh                    5820 non-null   object        
 6   Nguồn                        5820 non-null   object        
 7   Mã vận đơn                   5820 non-null   object        
 8   Tình trạng gói hàng          5820 non-null   object        
 9   Trạng thái đối tác           5819 non-null   object        
 10  Lý do hủy đơn                0 non-null      float64       
 11  Ngày đóng gói                5820 non-null 

In [ ]:
vd_1.shape

(7944, 45)

In [ ]:
#Xóa cột STT
vd_1.drop(columns = ['STT'], inplace = True)

**Xử lí các giá trị null**

In [ ]:
vd_1.isna().sum()

Mã đơn hàng                    2124
Ghi chú đơn hàng               7378
Tags đơn hàng                  7917
Nhân viên tạo đơn              2124
Chi nhánh                      2124
Nguồn                          2124
Mã vận đơn                     2124
Tình trạng gói hàng            2124
Trạng thái đối tác             2125
Lý do hủy đơn                  7944
Ngày đóng gói                  2124
Ngày hẹn giao                  7944
Ngày xuất kho                  2124
Ngày giao hàng                 2829
Đối tác giao hàng              2124
Dịch vụ giao hàng              2125
Khối lượng(g)                  2124
Kích thước(DxRxC)              2124
Tên người nhận                 2124
SĐT người nhận                 2124
Địa chỉ giao hàng              2124
Tỉnh/Thành                     2124
Quận/Huyện                     2124
Phường xã                      2124
Trạng thái đối soát            2124
Tiền khách phải trả cho đơn    2124
Khách hàng đã trả              2911
Hình thức thanh toán        

Nhận thấy có nhiều cột cùng có 2124 giá trị null, kiểm tra vùng dữ liệu này.

In [ ]:
#Kiểm tra vùng dữ liệu khi Mã đơn hàng null
df = vd_1[vd_1['Mã đơn hàng'].isna()]
df.isna().mean()

Mã đơn hàng                    1.000000
Ghi chú đơn hàng               1.000000
Tags đơn hàng                  1.000000
Nhân viên tạo đơn              1.000000
Chi nhánh                      1.000000
Nguồn                          1.000000
Mã vận đơn                     1.000000
Tình trạng gói hàng            1.000000
Trạng thái đối tác             1.000000
Lý do hủy đơn                  1.000000
Ngày đóng gói                  1.000000
Ngày hẹn giao                  1.000000
Ngày xuất kho                  1.000000
Ngày giao hàng                 1.000000
Đối tác giao hàng              1.000000
Dịch vụ giao hàng              1.000000
Khối lượng(g)                  1.000000
Kích thước(DxRxC)              1.000000
Tên người nhận                 1.000000
SĐT người nhận                 1.000000
Địa chỉ giao hàng              1.000000
Tỉnh/Thành                     1.000000
Quận/Huyện                     1.000000
Phường xã                      1.000000
Trạng thái đối soát            1.000000


Gần như tất cả các cột ở vùng dữ liệu này đều chứa 100% giá trị là giá trị null. Những cột có thông tin là những cột liên quan đến sản phẩm trong đơn hàng, còn giá trị của các cột còn lại đều là null. Quyết định xóa những cột này vì đây có thể là lỗi khi nhập liệu.

In [ ]:
#Xóa vùng dữ liệu trên
vd_1 = vd_1.drop(vd_1[vd_1['Mã đơn hàng'].isna()].index)

*Xóa một số cột* gần như hoặc tất cả bị null toàn bộ: Ghi chú đơn hàng, Tags đơn hàng, Lý do hủy đơn, Ngày hẹn giao, Ghi chú sản phẩm, Serial.

In [ ]:
#Xóa các cột
columns_to_drop = ['Ghi chú đơn hàng','Tags đơn hàng','Lý do hủy đơn','Ngày hẹn giao','Ghi chú sản phẩm','Serial']
vd_1.drop(columns = columns_to_drop, inplace = True)

In [ ]:
pd.set_option('display.max_columns', None)
vd_1.head()

STT Mã đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn     Mã vận đơn  \
0  1.0    SON25144  Nguyễn Thị Huyền  Chi nhánh mặc định   Web       G87L96F3   
2  2.0    SON24883  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web       G87L9VTM   
4  3.0    SON24943  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web  NSAPOAAAAZPQF   
5  4.0    SON25048  Đỗ Thị Hồng Ngọc  Chi nhánh mặc định   Web  NSAPOAAAAZPQE   
6  5.0    SON25126  Kiều Phương Thảo  Chi nhánh mặc định   Web  NSAPOAAAAZPQD   

  Tình trạng gói hàng Trạng thái đối tác       Ngày đóng gói  \
0        Đã giao hàng    Giao thành công 2020-09-16 12:49:54   
2        Đã giao hàng    Giao thành công 2020-09-16 12:49:13   
4        Đã giao hàng    Giao thành công 2020-09-15 19:56:20   
5        Đã giao hàng    Giao thành công 2020-09-15 19:55:54   
6        Đã giao hàng    Giao thành công 2020-09-15 19:55:30   

        Ngày xuất kho      Ngày giao hàng Đối tác giao hàng  \
0 2020-09-17 11:05:38 2020-09-19 10:20:38      Sapo Express   
2 2020-09-17 11:05:48 2020-09-21 12:18:59      Sapo Express   
4 2020-09-15 19:58:37 2020-09-19 12:54:17      Sapo Express   
5 2020-09-15 19:58:43 2020-09-18 14:46:31      Sapo Express   
6 2020-09-15 19:58:49 2020-09-18 11:03:22      Sapo Express   

     Dịch vụ giao hàng  Khối lượng(g)     Kích thước(DxRxC)    Tên người nhận  \
0                Chuẩn          500.0  10.000x10.000x10.000      Mẹ Bống Trần   
2                Chuẩn          500.0  10.000x10.000x10.000           Nhan Le   
4  (Ninja Van - Chuẩn)         1000.0  10.000x10.000x10.000    Vũ Thị Thúy Hà   
5  (Ninja Van - Chuẩn)         1000.0  10.000x10.000x10.000           Linh An   
6  (Ninja Van - Chuẩn)         1000.0  10.000x10.000x10.000  Nguyễn Mộng Hiền   

  SĐT người nhận                            Địa chỉ giao hàng      Tỉnh/Thành  \
0     0389913230                                    thôn Ngọc        Hưng Yên   
2     0907867931         trường mn Phú Thuận đường Phạm Thị B  TP Hồ Chí Minh   
4     0916522265  Bệnh viện phụ sản trung ương Triệu Quốc Đạt          Hà Nội   
5     0977703188                  Số 56, phố Trịnh Ngọc Điệt,       Thanh Hóa   
6     0905634664                           36 Nguyễn Công Trứ       Quảng Nam   

            Quận/Huyện         Phường xã Trạng thái đối soát  \
0        Huyện Văn Lâm        Xã Lạc Đạo         Đã đối soát   
2               Quận 7  Phường Phú Thuận         Đã đối soát   
4       Quận Hoàn Kiếm  Phường Hàng Bông         Đã đối soát   
5  Thành phố Thanh Hóa    Phường An Hưng         Đã đối soát   
6     Thành phố Tam Kỳ  Phường Tân Thạnh         Đã đối soát   

   Tiền khách phải trả cho đơn  Khách hàng đã trả Hình thức thanh toán  \
0                     499000.0           499000.0                  COD   
2                     499000.0           499000.0                  COD   
4                     780000.0           780000.0                  COD   
5                     780000.0           780000.0                  COD   
6                     779000.0           779000.0                  COD   

   Tổng tiền thu hộ  Phí vận chuyển Người trả phí  Phí trả đối tác  \
0          499000.0             0.0      Shop trả          26600.0   
2          499000.0             0.0      Shop trả          26600.0   
4          780000.0             0.0      Shop trả          22000.0   
5          780000.0             0.0      Shop trả          22000.0   
6          779000.0             0.0      Shop trả          22000.0   

                           Ghi chú đơn giao Mã sản phẩm  \
0                              Cho thử hàng          ph   
2                              Cho thử hàng     p2chung   
4  Cho thử hàng | Giao trong giờ hành chính   combokid2   
5                              Cho thử hàng  comboteen2   
6                              Cho thử hàng  comboteen2   

                                        Tên sản phẩm  Số lượng Đơn vị tính  \
0                                   Sách TeenUp - PH         1         NaN   
2               

***Cột 'Đơn vị tính'*** có rất nhiều giá trị null, kiểm tra xem có nên loại bỏ cột này không.

In [ ]:
#Các giá trị duy nhất trong cột
vd_1['Đơn vị tính'].unique()

array([nan, 'Bộ'], dtype=object)

Cột chỉ có giá trị 'Bộ' hoặc là null, không có nhiều giá trị khi phân tích, vì vậy xóa cột này.

In [ ]:
vd_1.drop(columns = 'Đơn vị tính', inplace = True)

***Cột 'Trạng thái đối tác'*** có 1 giá trị null, thực hiện xóa bản ghi này.

In [ ]:
vd_1 = vd_1.drop(vd_1[vd_1['Trạng thái đối tác'].isna()].index)

Kiểm tra vùng dữ liệu khi 'Ngày giao hàng' bị null.

In [ ]:
df = vd_1[vd_1['Ngày giao hàng'].isna()]
pd.set_option('display.max_columns', None)
df.head(10)

Mã đơn hàng Nhân viên tạo đơn           Chi nhánh Nguồn     Mã vận đơn  \
11     SON25391     Trần Thị Thúy  Chi nhánh mặc định   Web  NSAPOAAAAZPPX   
12     SON25304     Trần Thị Thúy  Chi nhánh mặc định   Web  NSAPOAAAAZPPW   
21     SON25749     Trần Thị Thúy  Chi nhánh mặc định   Web       G87Q4FUC   
47     SON25791     Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBURA   
75     SON25833     Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBG7W   
83     SON25818     Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBF3W   
84     SON25825     Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBF3P   
97     SON25345   Huyền Nguyễn CS  Chi nhánh mặc định   Web       G8FEEHPQ   
99     SON25326        Sáng Trịnh  Chi nhánh mặc định   Web       G8FEEPBD   
119    SON25360     Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEEDB7   

    Tình trạng gói hàng Trạng thái đối tác       Ngày đóng gói  \
11     Hủy giao-đã nhận     Trả thành công 2020-09-15 19:08:50   
12     Hủy giao-đã nhận        Chờ lấy lại 2020-09-15 19:08:35   
21     Hủy giao-đã nhận     Trả thành công 2020-09-15 17:30:05   
47     Hủy giao-đã nhận     Trả thành công 2020-09-15 16:37:40   
75     Hủy giao-đã nhận     Trả thành công 2020-09-15 16:36:05   
83     Hủy giao-đã nhận     Trả thành công 2020-09-15 16:34:24   
84     Hủy giao-đã nhận     Trả thành công 2020-09-15 16:34:24   
97     Hủy giao-đã nhận     Trả thành công 2020-09-15 11:46:07   
99     Hủy giao-đã nhận     Trả thành công 2020-09-15 11:37:05   
119    Hủy giao-đã nhận     Trả thành công 2020-09-15 11:35:57   

          Ngày xuất kho Ngày giao hàng Đối tác giao hàng  \
11  2020-09-15 19:39:08            NaT      Sapo Express   
12  2020-09-15 19:39:28            NaT      Sapo Express   
21  2020-09-15 17:34:49            NaT      Sapo Express   
47  2020-09-15 17:03:32            NaT      Sapo Express   
75  2020-09-15 17:07:52            NaT      Sapo Express   
83  2020-09-15 17:23:28            NaT      Sapo Express   
84  2020-09-15 17:23:34            NaT      Sapo Express   
97  2020-09-15 13:49:58            NaT      Sapo Express   
99  2020-09-15 13:50:20            NaT      Sapo Express   
119 2020-09-15 13:52:28            NaT      Sapo Express   

         Dịch vụ giao hàng  Khối lượng(g)     Kích thước(DxRxC)  \
11                   Chuẩn          500.0  10.000x10.000x10.000   
12                   Chuẩn          500.0  10.000x10.000x10.000   
21   (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
47   (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
75   (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
83   (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
84   (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
97   (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
99   (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   
119  (GHN Express - Chuẩn)          500.0  10.000x10.000x10.000   

         Tên người nhận SĐT người nhận  \
11      Nguyễn thị thúy     0345269989   
12      Nguyễn thị thúy     0345269989   
21        Phung duc nam     0918736749   
47         Nguyễn ly na     0915992835   
75         Trang Nguyen     0983235102   
83                Tigon     0947859638   
84       nguyễn thị thu     0384576286   
97            Chị Hường     0979080354   
99   Trần Thị Diệu Oanh     0842078178   
119       Thiennga tran     0935282984   

                                     Địa chỉ giao hàng      Tỉnh/Thành  \
11   Phường Thọ Quang Quận Sơn Trà Đà Nẵng Việt Nam...         Đà Nẵng   
12   Phường Thọ Quang Quận Sơn Trà Đà Nẵng Việt Nam...         Đà Nẵng   
21   Phường Võ Thị Sáu Quận 3 Hồ Chí Minh Việt Nam ...  TP Hồ Chí Minh   
47   Phường Nam Lý Thành phố Đồng Hới Quảng Bình Vi...      Quảng Bình   
75   Phường Hoàng Văn Thụ Quận Hoàng Mai Hà Nội Việ...          Hà Nội   
83   Phường 9 Quận Phú Nhuận Hồ Chí Minh Việt Nam 1...  TP Hồ Chí Minh   
84   Xã Tam Hồng Huyện Yên Lạc Vĩnh Ph

Nhận thấy vùng vùng dữ liệu này có điểm chung là cột 'Tình trạng gói hàng' đều có giá trị 'Hủy giao-đã nhận'. Thử kiểm tra các giá trị xuất hiện trong cột.

In [ ]:
#Kiểm tra cột các giá trị unique của cột 'Tình trạng gói hàng'
df['Tình trạng gói hàng'].unique()

array(['Hủy giao-đã nhận', 'Đang giao hàng', 'Chờ giao lại'], dtype=object)

Các giá trị trong cột gồm: 'Hủy giao-đã nhận', 'Đang giao hàng', 'Chờ giao lại'. Như vậy có thể giải thích lý do cột 'Ngày giao hàng' bị null là do những đơn hàng này đã bị hủy hoặc chưa tới được tay khách hàng, điều này là hoàn toàn hợp lý.

Kiểm tra vùng dữ liệu khi 'Khách hàng đã trả' null.

In [ ]:
df = vd_1[vd_1['Khách hàng đã trả'].isna()]
df.head(20)

Mã đơn hàng    Nhân viên tạo đơn           Chi nhánh Nguồn     Mã vận đơn  \
10     SON25590  Đinh Thục Hoàng Anh  Chi nhánh mặc định   Web       G87QKEGR   
11     SON25391        Trần Thị Thúy  Chi nhánh mặc định   Web  NSAPOAAAAZPPX   
12     SON25304        Trần Thị Thúy  Chi nhánh mặc định   Web  NSAPOAAAAZPPW   
13     SON25136        Đỗ Khánh Linh  Chi nhánh mặc định   Web       G87QKM8B   
14     SON24843   Trương Minh Nguyệt  Chi nhánh mặc định   Web       G87QKM3Q   
21     SON25749        Trần Thị Thúy  Chi nhánh mặc định   Web       G87Q4FUC   
47     SON25791        Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBURA   
75     SON25833        Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBG7W   
83     SON25818        Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBF3W   
84     SON25825        Trần Thị Thúy  Chi nhánh mặc định   Web       G87QBF3P   
97     SON25345      Huyền Nguyễn CS  Chi nhánh mặc định   Web       G8FEEHPQ   
99     SON25326           Sáng Trịnh  Chi nhánh mặc định   Web       G8FEEPBD   
119    SON25360        Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEEDB7   
120    SON25364        Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEEDPH   
121    SON25368        Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEEDPK   
123    SON25371        Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEEDDR   
124    SON25384        Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEEDDK   
125    SON25400        Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEEDD9   
127    SON25445        Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEEDN8   
131    SON25513        Trần Thị Thúy  Chi nhánh mặc định   Web       G8FEED6U   

    Tình trạng gói hàng Trạng thái đối tác       Ngày đóng gói  \
10         Đã giao hàng    Giao thành công 2020-09-15 19:09:06   
11     Hủy giao-đã nhận     Trả thành công 2020-09-15 19:08:50   
12     Hủy giao-đã nhận        Chờ lấy lại 2020-09-15 19:08:35   
13         Đã giao hàng    Giao thành công 2020-09-15 19:08:12   
14         Đã giao hàng    Giao thành công 2020-09-15 19:07:00   
21     Hủy giao-đã nhận     Trả thành công 2020-09-15 17:30:05   
47     Hủy giao-đã nhận     Trả thành công 2020-09-15 16:37:40   
75     Hủy giao-đã nhận     Trả thành công 2020-09-15 16:36:05   
83     Hủy giao-đã nhận     Trả thành công 2020-09-15 16:34:24   
84     Hủy giao-đã nhận     Trả thành công 2020-09-15 16:34:24   
97     Hủy giao-đã nhận     Trả thành công 2020-09-15 11:46:07   
99     Hủy giao-đã nhận     Trả thành công 2020-09-15 11:37:05   
119    Hủy giao-đã nhận     Trả thành công 2020-09-15 11:35:57   
120    Hủy giao-đã nhận     Trả thành công 2020-09-15 11:35:56   
121    Hủy giao-đã nhận     Trả thành công 2020-09-15 11:35:55   
123    Hủy giao-đã nhận     Trả thành công 2020-09-15 11:35:54   
124    Hủy giao-đã nhận     Trả thành công 2020-09-15 11:35:53   
125    Hủy giao-đã nhận     Trả thành công 2020-09-15 11:35:53   
127    Hủy giao-đã nhận     Trả thành công 2020-09-15 11:35:52   
131    Hủy giao-đã nhận       Chờ lấy hàng 2020-09-15 11:35:49   

          Ngày xuất kho      Ngày giao hàng Đối tác giao hàng  \
10  2020-09-15 19:42:26 2020-09-19 14:12:25      Sapo Express   
11  2020-09-15 19:39:08                 NaT      Sapo Express   
12  2020-09-15 19:39:28                 NaT      Sapo Express   
13  2020-09-15 19:41:49 2020-09-18 20:12:15      Sapo Express   
14  2020-09-15 19:41:17 2020-09-17 11:37:03      Sapo Express   
21  2020-09-15 17:34:49                 NaT      Sapo Express   
47  2020-09-15 17:03:32                 NaT      Sapo Express   
75  2020-09-15 17:07:52                 NaT      Sapo Express   
83  2020-09-15 17:23:28                 NaT      Sapo Express   
84  2020-09-15 17:23:34                 NaT      Sapo Express   
97  2020-09-15 13:49:58                 NaT      Sapo Express   
99  2020-09-15 13:50:20                 NaT      Sapo Express   
119 2020-09-15 13:52:28                 NaT      Sapo Express   
120 2

Quan sát thấy các giá trị trong cột 'Hình thức thanh toán' đều null, điều này là hợp lý do không có thông tin về việc thanh toán của khách hàng.

**Kiểm tra giá trị trong các Categorical Column**

Kiểm tra các cột: Chi nhánh, Nguồn, Tình trạng gói hàng, Trạng thái đối tác, Đối tác giao hàng, Trạng thái đối soát, Người trả phí.

In [ ]:
#Kiểm tra giá trị riêng biệt của từng cột
df = vd_1[['Chi nhánh','Nguồn','Tình trạng gói hàng','Trạng thái đối tác','Đối tác giao hàng','Trạng thái đối soát','Người trả phí']]
for col in df.columns:
  unique_value = {col: df[col].unique()}
  print(unique_value)
  print('------------------------------------------')

{'Chi nhánh': array(['Chi nhánh mặc định'], dtype=object)}
------------------------------------------
{'Nguồn': array(['Web', 'Other', 'Lazada'], dtype=object)}
------------------------------------------
{'Tình trạng gói hàng': array(['Đã giao hàng', 'Hủy giao-đã nhận', 'Đang giao hàng',
       'Chờ giao lại'], dtype=object)}
------------------------------------------
{'Trạng thái đối tác': array(['Giao thành công', 'Trả thành công', 'Chờ lấy lại', 'Chờ lấy hàng',
       'Chờ trả hàng', 'Giao hàng thành công', 'Đang lấy hàng',
       'Đã giao hàng', 'Đang giao hàng', 'Thất lạc'], dtype=object)}
------------------------------------------
{'Đối tác giao hàng': array(['Sapo Express', 'GrabExpress', 'Ahamove'], dtype=object)}
------------------------------------------
{'Trạng thái đối soát': array(['Đã đối soát', 'Chưa đối soát'], dtype=object)}
------------------------------------------
{'Người trả phí': array(['Shop trả'], dtype=object)}
------------------------------------------


Các giá trị trong những cột này đều hợp lý.

Trong cột Trạng thái đối tác, chuyển giá trị 'Giao hàng thành công' và 'Đã giao hàng' sang 'Giao thành công' do chúng mang ý nghĩa như nhau.

In [ ]:
#Trạng thái đối tác
#Chuyển 'Giao hàng thành công' và 'Đã giao hàng' sang 'Giao thành công'
vd_1['Trạng thái đối tác'] = vd_1['Trạng thái đối tác'].replace(['Giao hàng thành công','Đã giao hàng'], 'Giao thành công')

In [ ]:
vd_1 = vd_1.drop_duplicates()

In [ ]:
vd_1.to_excel('Vận đơn_cleaned.xlsx',index = False)

# **Phần 2: Xây dựng báo cáo dựa trên dữ liệu có sẵn**

## **Câu 1**: CEO đang cần dựng một báo cáo về của tình hình vận hành doanh nghiệp trong Q3/2020. Theo bạn, các chỉ số nào là quan trọng và cần đưa vào báo cáo?

**Các chỉ số cần đưa vào báo cáo**

1- Đánh giá hoạt động Marketing

- Doanh thu: bằng tổng của cột Paid Revenue, đánh giá hiệu quả kinh doanh
- Lợi nhuận: bằng Paid Revenue - chi phí Marketing, đánh giá hiệu quả kinh doanh
- Chỉ số ROI: tỉ lệ giữa lợi nhuận và chi phí Marketing, đánh giá lợi ích thu được của công ty trong hoạt động Marketing
- Số lượng khách hàng tiềm năng (Lead), số lượng đơn hàng
- Chỉ số Conversion Rate (tỷ lệ chuyển đổi): tỉ lệ giữa số lượng khách hàng thực sự (đã mua hàng và thanh toán đầy đủ, là tất cả các khách hàng có thông tin trong cột 'Closed Date' tại bảng Sales) và tổng số Lead. Chỉ số này giúp đo lường hiệu quả cũng như theo dõi hiệu suất Marketing.

2- Đánh giá hoạt động bán hàng
- Doanh thu
- Số đơn hàng được hoàn thành: tổng số đơn hàng được thanh toán
- Thời gian trung bình hoàn thành một đơn hàng

3- Hoạt động vận đơn
- Tỷ lệ giao thành công: tỷ lệ giữa số đơn giao thành công và tổng số đơn được giao
- Tỷ lệ hoàn trả: tỷ lệ giữa số đơn hoàn trả và tổng số đơn được giao
- Thời gian giao hàng
- Chi phí trả đối tác


**Bổ sung một số cột để làm báo cáo**

1- Thời gian giao hàng



In [ ]:
vd_1  = vd.copy()

In [ ]:
vd_1['Thời gian giao hàng'] = vd_1['Ngày giao hàng'] - vd_1['Ngày xuất kho']

In [ ]:
vd_1['Thời gian giao hàng']

0      1 days 23:15:00
1      4 days 01:13:11
2      3 days 16:55:40
3      2 days 18:47:48
4      2 days 15:04:33
             ...      
5814   1 days 17:47:29
5815   2 days 00:06:36
5816   0 days 20:45:57
5817               NaT
5818   0 days 18:01:57
Name: Thời gian giao hàng, Length: 5819, dtype: timedelta64[ns]

In [ ]:
from datetime import timedelta
vd_1['Thời gian giao hàng'] = vd_1['Thời gian giao hàng'].dt.total_seconds() / 3600
vd_1['Thời gian giao hàng']

0       47.250000
1       97.219722
2       88.927778
3       66.796667
4       63.075833
          ...    
5814    41.791389
5815    48.110000
5816    20.765833
5817          NaN
5818    18.032500
Name: Thời gian giao hàng, Length: 5819, dtype: float64

In [ ]:
vd_1['Thời gian giao hàng'] = round(vd_1['Thời gian giao hàng'], 0)
vd_1['Thời gian giao hàng']

0       47.0
1       97.0
2       89.0
3       67.0
4       63.0
        ... 
5814    42.0
5815    48.0
5816    21.0
5817     NaN
5818    18.0
Name: Thời gian giao hàng, Length: 5819, dtype: float64

In [ ]:
vd_1 = vd_1.rename(columns={'Thời gian giao hàng': 'Thời gian giao hàng (giờ)'})

2- Miền

Địa điểm giao hàng theo miền

In [ ]:
domain_mapping = {
    'Miền Bắc': ['Phú Thọ','Hải Dương','Bắc Ninh','Bắc Giang','Cao Bằng','Điện Biên','Hải Phòng','Sơn La','Hòa Bình','Vĩnh Phúc','Hà Nội','Lạng Sơn','Bắc Kạn','Lào Cai','Quảng Ninh','Hà Nam','Hà Giang','Tuyên Quang','Lai Châu','Thái Nguyên','Ninh Bình','Nam Định','Hưng Yên','Yên Bái','Thái Bình'],
    'Miền Trung': ['Quảng Ngãi','Lâm Đồng','Khánh Hòa','Thanh Hóa','Đà Nẵng','Đắk Nông','Quảng Nam','Nghệ An','Quảng Bình','Bình Định','Kon Tum','Phú Yên','Đắk Lắk','Quảng Trị','Hà Tĩnh','Gia Lai','Thừa Thiên Huế'],
    'Miền Nam': ['Trà Vinh','Bình Thuận','Hậu Giang','Bạc Liêu','Ninh Thuận','Vĩnh Long','An Giang','Đồng Nai','Tây Ninh','Tiền Giang','Bà Rịa-Vũng Tàu','Kiên Giang','Bến Tre','Bình Dương','TP Hồ Chí Minh','Bình Phước','Đồng Tháp','Cần Thơ','Long An','Cà Mau','Sóc Trăng']
}
vd_1['Miền'] = vd_1['Tỉnh/Thành'].apply(lambda x: next((key for key, value in domain_mapping.items() if x in value), None))

In [ ]:
vd_1[['Miền','Tỉnh/Thành']]

Miền      Tỉnh/Thành
0       Miền Bắc        Hưng Yên
1       Miền Nam  TP Hồ Chí Minh
2       Miền Bắc          Hà Nội
3     Miền Trung       Thanh Hóa
4     Miền Trung       Quảng Nam
...          ...             ...
5814  Miền Trung         Nghệ An
5815    Miền Nam  TP Hồ Chí Minh
5816    Miền Bắc       Hải Phòng
5817    Miền Bắc        Nam Định
5818    Miền Bắc         Phú Thọ

[5819 rows x 2 columns]